src/compiler/checker.ts

```diff
@@ -8977,16 +8978,35 @@ namespace ts {
                 }
             }
 
-            function typeRelatedToEachType(source: Type, target: UnionOrIntersectionType, reportErrors: boolean): Ternary {
+            function typeRelatedToEachType(source: Type, target: IntersectionType, reportErrors: boolean): Ternary {
                 let result = Ternary.True;
                 const targetTypes = target.types;
+                const saveDynamicDisableWeakTypeErrors = dynamicDisableWeakTypeErrors;
+                dynamicDisableWeakTypeErrors = true;
                 for (const targetType of targetTypes) {
                     const related = isRelatedTo(source, targetType, reportErrors);
                     if (!related) {
+                        dynamicDisableWeakTypeErrors = saveDynamicDisableWeakTypeErrors;
                         return Ternary.False;
                     }
                     result &= related;
                 }
+                dynamicDisableWeakTypeErrors = saveDynamicDisableWeakTypeErrors;
+                if (source !== globalObjectType && getPropertiesOfType(source).length > 0 && every(target.types, isWeak)) {

```

Comment:
I refactored this into a separate function and a named boolean. I think the two new names should serve as adequate documentation.